File: cmd/code/replaceNonCoprimes.go
package code

// helper of find gcd
func gcd(a int, b int) int {
	for b != 0 {
		var temp int = b
		b = a % b
		a = temp
	}

	return  a
}

// helper to find lcm
func lcm(a int, b int) int {
	return (a * b) / gcd(a, b)
}

// main function
func replaceNonCoprimes(nums []int) []int {
	stack := []int{}

	for _,num := range nums {
		for len(stack) > 0 {
			top := stack[len(stack) - 1]
			g := gcd(top, num)

			if g > 1 {
				num = lcm(top, num)
				stack = stack[:len(stack) - 1]
			} else {
				break
			}
		}

		stack = append(stack, num)
	}

	return  stack
}

File: cmd/code/replaceNonCoprimes.js

// helper ot find gcd
function gcd ( a, b )
{
    while ( b !== 0 )
    {
        let temp = b
        b = a % b
        a = temp
    }

    return a
}

// helper to find lcm
function lcm ( a, b )
{
    return ( a * b ) / gcd( a, b );
}

// main function
function replaceNonCoprimes ( nums )
{
    const stack = []
    
    for ( let num of nums )
    {
        while ( stack.length > 0 )
        {
            const top = stack[ stack.length - 1 ];
            const g = gcd( top, num )
            if ( g > 1 )
            {
                num = lcm( top, num )
                stack.pop()
            }
            else
            {
                break;
            }
        }

        stack.push(num)
    }

    return stack
}

console.log(replaceNonCoprimes([6, 4, 3, 2, 7, 6, 2])); // [12, 7, 6]
console.log(replaceNonCoprimes([2, 2, 1, 1, 3, 3, 3])); // [2, 1, 1, 3]